ESA_JWST-doc-update: actual examples in JWST module

jespinosaar opened the pull request:
Dear astroquery team:

This PR aims to update the documentation in ESA JWST module with more actual examples. I hope it is simple to merge. Thanks in advance for your support!

cc @esdc-esac-esa-int

keflavich commented on docs/esa/jwst/jwst.rst:
```diff
@@ -104,23 +104,22 @@ service degradation.
   >>> width=u.Quantity(5, u.deg)
   >>> height=u.Quantity(5, u.deg)
   >>> r=Jwst.query_region(coordinate=coord, width=width, height=height)
-  >>> r
```
was this deletion intentional (and many similar ones later)?  I believe you need this to get the printout below

jespinosaar replied:
Yes, I did not see it in other modules, that is why I removed them. But I can include them again.

jespinosaar replied:
Ok, done!

bsipocz replied:
I think it's a bit of mixed bag for the other modules. In general we tend to show the result of a query, but in some cases it's not very deterministic, so then we don't, or skip their doctest. 

Either way, I think it's better to avoid single letter variables everywhere, especially in the documentation and call everything something meaningful (just as you did with `width` and `height`, so what about using `result` or something similar?)

bsipocz commented:
No, we don't do changelog for docs or any non user facing internal changes.

jespinosaar commented:
Then, if there are not any other comments, maybe we can merge this PR.

eerovaher commented on docs/esa/jwst/jwst.rst:
```diff
@@ -225,69 +224,67 @@ To query the data products associated with a certain Observation ID
 
   >>> from astroquery.esa.jwst import Jwst
-  >>> product_list=Jwst.get_product_list(observation_id='jw00777011001_02104_00001_nrcblong')
+  >>> product_list=Jwst.get_product_list(observation_id='jw01063107001_02101_00013_nrca3')
```
It would be better if the code examples followed [PEP 8](https://peps.python.org/pep-0008/). In particular [`=` in variable assignment should be surrounded with spaces](https://peps.python.org/pep-0008/#other-recommendations). I'm highlighting it only here, but there are more instances of this in the file.
```suggestion
  >>> product_list = Jwst.get_product_list(observation_id='jw01063107001_02101_00013_nrca3')
```

eerovaher commented on docs/esa/jwst/jwst.rst:
```diff
@@ -328,20 +330,16 @@ To load only table names (TAP+ capability)
   >>> for table in (tables):
```
```suggestion
  >>> for table in tables:
```

eerovaher commented on docs/esa/jwst/jwst.rst:
```diff
@@ -423,97 +406,70 @@ Query without saving results in a file:
   >>> from astroquery.esa.jwst import Jwst
   >>>
   >>> job=Jwst.launch_job("SELECT TOP 100 \
-  >>> instrument_name, observationuri, planeid, calibrationlevel, \
+  >>> instrument_name, proposal_id, calibrationlevel, \
   >>> dataproducttype \
   >>> FROM jwst.main ORDER BY instrument_name, observationuri")
```
Instead of using backslashes PEP 8 recommends implied line continuation inside parentheses:
```python
job = Jwst.launch_job("SELECT TOP 100 "
    "instrument_name, proposal_id, calibrationlevel "
    "dataproducttype "
    "FROM jwst.main ORDER BY instrument_name, observationuri")
```
But triple quotes might also be acceptable:
```python
job = Jwst.launch_job("""SELECT TOP 100 
    instrument_name, proposal_id, calibrationlevel 
    dataproducttype 
    FROM jwst.main ORDER BY instrument_name, observationuri""")
```

jespinosaar commented on docs/esa/jwst/jwst.rst:
```diff
@@ -389,21 +380,13 @@ Once a table is loaded, columns can be inspected
   >>> for column in (table.columns):
   >>>   print(column.name)
```
Ok, changes applied.

jespinosaar commented on docs/esa/jwst/jwst.rst:
```diff
@@ -324,86 +328,66 @@ To load only table names (TAP+ capability)
-  type
-  typecode
+  >>> table = Jwst.load_table('jwst.main')
+  >>> print(*(column.name for column in table.columns), sep="\n")
```
That line is not working for me:
AttributeError: 'TapTableMeta' object has no attribute 'colnames'

eerovaher replied:
I assumed that `table` is an instance of `astropy.table.Table`, but apparently it is an instance of `astroquery.utils.tap.model.taptable.TapTableMeta`. It is not obvious to me why the `TapTableMeta` class exists or if it has to exist then why wouldn't it inherit from `Table`, but even if that is something that should be addressed, it would be beyond the scope of this pull request.

jespinosaar replied:
I agree with that, so I will leave this line as it is.

jespinosaar commented:
Thanks for your feedback @eerovaher , @keflavich , @bsipocz ! everything is now solved so, if you agree, we can merge this PR.

jespinosaar commented:
Dear all, can we close this PR?

bsipocz commented:
> Dear all, can we close this PR?

I'm looking into this, I suppose vacation season means we're away from the keyboard from time to time (I certainly was away having some ⛰️ refill in the past few days) 😄

bsipocz reviewed (commented):
I haven't yet gone into any nitpicking details but first wonder whether  you would have the capacity to remove the `doctest-skip-all` directive from the top to start running doctest on these code snippets?

After that removal you would need to add the ``.. doctest-remote-data:`` for each and every code block, but hopefully that is all, as the examples are very freshly made I suppose they will pass running them as tests.

bsipocz commented on docs/esa/jwst/jwst.rst:
```diff
@@ -225,68 +224,68 @@ To query the data products associated with a certain Observation ID
-  >>>     print("filename: %s" % (row['filename']))
+  >>> product_list = Jwst.get_product_list(observation_id='jw01063107001_02101_00013_nrca3')
+  >>> print(product_list['filename'])
+
```
doctest won't like empty lines like this and expects the output to be placed straight after the code lines.

jespinosaar replied:
Fixed

bsipocz commented on docs/esa/jwst/jwst.rst:
```diff
@@ -225,68 +224,68 @@ To query the data products associated with a certain Observation ID
-  >>> query="select a.artifactid, a.uri from jwst.artifact a, jwst.plane p where p.planeid=a.planeid and p.obsid='00000000-0000-0000-9c08-f5be8f3df805'"
-  >>> job=Jwst.launch_job(query, async_job=True)
-  >>> results=job.get_results()
+  >>> query = "select o.observationid, a.artifactid, a.filename from jwst.observation o join jwst.artifact a on a.obsid = o.obsid where o.proposal_id = '01166' and o.intent = 'science'"
```
maybe for `o` and `a` also use something a bit longer (obs and art? though those are not that much better either). Anyway, take this as my thinking out loud, I don't in fact have a strong preference just wonder what would work nicer.

Also, it is worth looking at the rendered docs and breaking long lines rather than having one very long in the horizontal  scroll

jespinosaar replied:
Fixed

bsipocz commented on docs/esa/jwst/jwst.rst:
```diff
@@ -422,98 +406,71 @@ Query without saving results in a file:
-  >>> instrument_name, observationuri, planeid, calibrationlevel, \
-  >>> dataproducttype \
-  >>> FROM jwst.main ORDER BY instrument_name, observationuri")
+  >>> job = Jwst.launch_job("SELECT TOP 100 "
+  >>>       "instrument_name, proposal_id, calibrationlevel, "
+  >>>       "dataproducttype "
+  >>>       "FROM jwst.main ORDER BY instrument_name, observationuri")
```
This won't work in this format, continued lines need the `...` instead of the `>>>` and I would also consider adding a proper indent to help reading.
```suggestion
  >>> job = Jwst.launch_job("SELECT TOP 100 "
  ...                       "instrument_name, proposal_id, calibrationlevel, "
  ...                       "dataproducttype "
  ...                       "FROM jwst.main ORDER BY instrument_name, observationuri")
```

jespinosaar replied:
Fixed

bsipocz commented on docs/esa/jwst/jwst.rst:
```diff
@@ -524,11 +481,11 @@ A table can be uploaded to the server in order to be used in a query.
-  >>> r=j.get_results()
+  >>> upload_resource = 'mytable.xml.gz'
+  >>> j = Jwst.launch_job(query="SELECT * from tap_upload.table_test",
+  >>>                   upload_resource=upload_resource,
```
same here, use `...` instead of `>>>` for continued lines. 

I would say this is the only issue I need to have fixed for merging

jespinosaar replied:
Fixed

bsipocz reviewed (changes requested):
I would say that fixing the `...` vs `>>>` issues are the only thing that needs to be fixed for merging, everything else is nitpicking, or thinking out loudly, or suggesting for future improvements.

I would love to have this module's narrative docs doctested (as said in the previous comment), but that can be done in a new follow-up PR. I think it's likely enough to change all the ``.. code-block:: python`` lines to `.. doctest-remote-data::`, and maybe some tweaking is needed for the outputs and formatting, but I don't expect much complications. Anyways, we can do this gradually, e.g. merging this PR and do opt out into testing in a new PR.

jespinosaar commented:
> > Dear all, can we close this PR?
> 
> I'm looking into this, I suppose vacation season means we're away from the keyboard from time to time (I certainly was away having some ⛰️ refill in the past few days) 😄

Dear @bsipocz , I am so sorry, I did not know you were away... Of course, do not worry if you are still on holiday, I will solve the issues in the meantime...

jespinosaar commented:
Ok, I think all the required changes have been applied.

bsipocz commented:
I went ahead and fixed the remaining issues, the main ones:

- some of the queries run into ValueError due to authentication issues. I skipped those for now, but some docs would be nice for it.
```
UNEXPECTED EXCEPTION: ValueError("Cannot retrieve products for observation jw01122001001_0210r_00001_nrs2: Error 403:\nPrivate file(s) requested: MAST token required for authentication.\n\nCannot process request: 'https://jwst.esac.esa.int/server/data' (req: Reqid: anonymous1656430293286, retrieval access: DIRECT, retrieval type: OBSERVATION, compression: null), for user: owner id: anonymous, roles: 0, session: C66DC70A30E7CCDDD6FE569302E2BFF9, ip: 73.109.49.174,  due to: Private file(s) requested: MAST token required for authentication.")
Traceback (most recent call last):
```

There are a lot of `INFO` messages coming from the module, I wonder whether those could be silenced, maybe by default. At the moment I've added them wherever they were issues.

bsipocz reviewed (approved):
Some examples have been temporarily skipped, otherwise, all of them pass the tests now.

bsipocz commented:
Thanks @jespinosaar!

jespinosaar commented:
> I went ahead and fixed the remaining issues, the main ones:
> 
>     * some of the queries run into ValueError due to authentication issues. I skipped those for now, but some docs would be nice for it.
> 
> 
> ```
> UNEXPECTED EXCEPTION: ValueError("Cannot retrieve products for observation jw01122001001_0210r_00001_nrs2: Error 403:\nPrivate file(s) requested: MAST token required for authentication.\n\nCannot process request: 'https://jwst.esac.esa.int/server/data' (req: Reqid: anonymous1656430293286, retrieval access: DIRECT, retrieval type: OBSERVATION, compression: null), for user: owner id: anonymous, roles: 0, session: C66DC70A30E7CCDDD6FE569302E2BFF9, ip: 73.109.49.174,  due to: Private file(s) requested: MAST token required for authentication.")
> Traceback (most recent call last):
> ```
> 
> There are a lot of `INFO` messages coming from the module, I wonder whether those could be silenced, maybe by default. At the moment I've added them wherever they were issues.

That is because we currently don't have data, there will be public data after the ERO release on July, 13th, and then I will update the examples with real download. Many thanks for looking into this!

jespinosaar commented:
Many thanks @bsipocz !!
